### ISSUE TITLE
完全情報手牌からcanonical structural wait ground truthを導出する

### ISSUE THREAD

**Issue by lisbun** [lisbun/lisjong#84]
## 背景

Issue #82 で `HandBelief` に、34基本牌種canonical axisを共有するstructural wait belief representationを追加した。

現在は以下をLevel 2 `HandBelief`として表現できる。

```text
wait_probability_raw                [34]
tanki_wait_probability_raw          [34]
shanpon_wait_probability_raw        [34]
kanchan_wait_probability_raw        [34]
penchan_wait_probability_raw        [34]
ryanmen_low_side_probability_raw    [34]
ryanmen_high_side_probability_raw   [34]
kokushi_wait_probability_raw        [34]
```

しかし、現時点ではこのtableへ**完全情報から正しい0/1 ground truthを生成する処理**は存在しない。

将来、

```text
完全情報牌譜
    ↓
exact wait ground truth
    ↓
training dataset
    ↓
hidden informationからwaitを予測
    ↓
HandBelief
```

という教師あり学習経路を構築するため、まず完全情報の手牌状態から決定論的にcanonical wait tableを生成できる純粋なbuilderを実装する。

## Goal

完全に判明している、

- concealed hand
- そのplayer自身の既知のmeld一式（chi / pon / open kan / added kan / concealed kanを含む）

を入力として、Issue #82 の契約に従う**Level 2 exact `HandBelief`**を決定論的に生成できるようにする。

出力するwait-related raw値はすべて、

```text
0
または
SCALE (= 8192)
```

とする。

つまり本Issueではprobabilistic predictionは行わず、**binary ground truth generator**を実装する。

## API方針

公開APIの具体的なfunction名・引数型は既存domain modelと整合させて決めるが、概念上は以下とする。

```python
exact_hand_belief_with_waits(
    concealed_hand,
    own_melds,
) -> HandBelief
```

入力には既存のcanonical `Tile` / `OwnHandState` / meld関連型を再利用し、RiichiEnv / RiichiLab固有型やwait生成専用の並行domain modelを新設しない。

既存のmeld modelがconcealed kanを含む必要な情報を表現できない場合は、別の並行型を即興で追加せず、blockingとして報告する。

返却値はIssue #82で固定した `HandBelief` とし、

```text
has_wait_belief == True
has_wait_mechanism_belief == True
```

となるLevel 2を返す。

既存 `exact_self_belief()` の責務を不用意に変更する必要はない。内部ロジック共有が自然ならrefactorしてよいが、既存behaviorは維持する。

## HandBelief内の責務境界

このbuilderが返す `HandBelief` では、tile marginalsとwait beliefの条件を明確に分ける。

```text
expected_count / red_five_probability
    = concealed handのみからexact生成

wait / wait mechanism channels
    = concealed hand + own meldsを条件に導出
```

meld内の牌を `expected_count_raw` へ加算してはならない。

`HandBelief.expected_count` は引き続き非公開concealed handのmarginalであり、wait beliefだけがknown own meldsを条件としたderived hand-state beliefである。

## 対象state: stable 13-equivalent hand

wait ground truthは、**打牌後などの「次の1牌で完成するか」を評価できるstable hand state**に対して生成する。

concealed handと完成済みmeldについて、structural equivalent countとして、

```text
len(concealed_tiles) + 3 * meld_count == 13
```

を要求する。

chi / pon / open kan / added kan / concealed kanはいずれも、4面子1雀頭を構成するうえでは完成済み1 meldなので、structural equivalent countではすべて `+3` と数える。

例:

```text
13 concealed + 0 meld
10 concealed + 1 meld
 7 concealed + 2 meld
 4 concealed + 3 meld
 1 concealed + 4 meld
```

14-equivalentのdrawn stateなどを暗黙に「非聴牌」としてall-zeroへ変換してはならない。

`OwnHandState`を入力に使う場合、`drawn_tile`を含むstateがこのstable 13-equivalent contractと矛盾するならfail-closedとする。builderがdrawn stateを暗黙にdiscard後stateへ変換してはならない。

## Structural countとphysical tile countを分離する

槓について、structural countとphysical tile countは別契約とする。

### Structural equivalent count

```text
chi / pon / any kan = 1 completed meld = 3-equivalent
```

したがってstable-state validationではkanも `+3` と数える。

### Physical tile count

5枚目candidate拒否など牌の物理制約を検証するときは、実際の牌枚数を数える。

```text
chi / pon = 3 physical tiles
kan       = 4 physical tiles
```

concealed kanもphysical countでは4枚である。

この2種類のcountを混同しない。

## Structural wait contract

Issue #82と同じく、waitは**structural completion**だけを意味する。

以下は考慮しない。

- yaku
- furiten
- ron / tsumo legality
- riichi
- 点数
- Policy action legality
- wall / dead wall
- 他playerの手牌
- 河に何枚見えているか
- global remaining inventory

つまり各34牌種 `t` について、

> 現在の手牌へ牌種`t`を1枚加えたとき、構造上有効な完成形が少なくとも1つ成立するか

を判定する。

成立すれば、

```text
wait_probability_raw[t] = SCALE
```

成立しなければ、

```text
wait_probability_raw[t] = 0
```

とする。

## Physical tile constraint

global remaining inventoryはwait判定に使用しない。

例えば1mが他playerの河等ですでに4枚見えていたとしても、入力手牌構造が1m待ちなら、

```text
wait[1m] = 1
```

でよい。

一方、そのplayer自身のconcealed hand + own meldsですでに同一基本牌種を4枚保持している場合、その牌種の5枚目candidateは物理的に成立しないためwaitとしない。

physical countは34基本牌種単位で行い、赤5 / 通常5を別牌種として扱わない。例えば赤5mを含めて5m系は合計最大4枚である。

また、入力state自体が同一基本牌種を5枚以上含むなど物理的に不正な場合は、candidate評価以前にfail-closedとする。

## Candidate enumeration

34基本牌種すべてについてcandidateを1種類ずつ評価する。

概念上、

```text
for t in all_34_tile_types:
    if adding t violates own physical tile count:
        continue

    completed_hand = concealed_hand + t

    evaluate:
        standard hand
        chiitoitsu
        kokushi

    record every mechanism that can complete the hand
```

とする。

赤5 / 通常5はwait axis上では区別しない。

iteration orderやinput tuple順によって結果が変化してはならない。

## Standard hand

通常手については、own meldsを完成済みmeldとして固定する。

own meld数を `m` とした場合、candidate追加後のconcealed部分から、

```text
(4 - m) melds + 1 pair
```

を構成できるすべての有効なdecompositionを考慮する。

own meldそのものを再分解しない。

chi / pon / open kan / added kan / concealed kanはいずれも完成済み1 meldとして扱う。

## Wait mechanism分類

candidate tile `t` を加えた完成形について、**追加された1枚`t`をどのgroupの完成に使用したと解釈できるか**によってmechanismを分類する。

physical copy identityは使用せず、

> candidate `t` をそのgroupから1枚取り除けば、元の13-equivalent concealed handに戻る

というexistentialな解釈で判定する。

### Tanki

candidateが完成形のpairを構成し、candidateを1枚取り除くとsingleton `t`になるdecompositionが存在すれば、

```text
tanki[t] = 1
```

とする。

### Shanpon

candidateがtripletを完成し、candidateを1枚取り除くとpairになるdecompositionが存在すれば、

```text
shanpon[t] = 1
```

とする。

### Kanchan

完成sequence `abc` についてcandidateが中央牌 `b` なら、

```text
a c -> b
```

として、

```text
kanchan[b] = 1
```

とする。

### Penchan

以下をpenchanとする。

```text
1 2 -> 3
8 9 -> 7
```

### Ryanmen low side

candidateがsequenceの低い側を完成し、penchanではない場合。

```text
2 3 -> 1
3 4 -> 2
...
7 8 -> 6
```

### Ryanmen high side

candidateがsequenceの高い側を完成し、penchanではない場合。

```text
2 3 -> 4
3 4 -> 5
...
7 8 -> 9
```

## すべてのdecompositionを考慮する

正解となるdecompositionを1つだけcanonical selectionしてはならない。

同一candidate `t`について、

```text
decomposition A -> tanki
decomposition B -> ryanmen
```

が成立するなら、

```text
wait[t] = 1
tanki[t] = 1
ryanmen_...[t] = 1
```

とする。

つまりground truthは**existential multi-label**である。

decomposition enumeration順、DFS順、input tile順などでlabelが変化してはならない。

## Chiitoitsu

七対子は**own meldsが0件の場合のみ**評価する。

「menzen / closed」という用語だけではconcealed kanを含む状態と混同し得るため、本Issueでは明示的に `meld_count == 0` を条件とする。

candidate追加後のhandが既存lisjongルール上の七対子完成形を満たす場合、

```text
wait[t] = 1
tanki[t] = 1
```

とする。

七対子専用wait channelは追加しない。

通常手としても完成可能な場合は、通常手側のmechanismも同時に立てる。

## Kokushi

国士無双も**own meldsが0件の場合のみ**評価する。

concealed kanを含め、meldが1件でも存在する場合は評価しない。

candidate追加後に国士無双のstructural completionが成立すれば、

```text
wait[t] = 1
kokushi[t] = 1
```

とする。

国士は`tanki`へ包含しない。

### 国士13面待ち

13幺九牌すべてについて、

```text
wait[t] = 1
kokushi[t] = 1
```

となる。

### 国士1種待ち

該当する1種類だけ1となる。

## Exact ground-truth consistency

本Issueが生成するLevel 2 exact ground truthでは、各tile `t`について必ず、

```text
wait[t] == OR(
    tanki[t],
    shanpon[t],
    kanchan[t],
    penchan[t],
    ryanmen_low_side[t],
    ryanmen_high_side[t],
    kokushi[t],
)
```

を満たす。

これはIssue #82の**probabilistic `HandBelief`一般**に課すconstructor constraintではない。

あくまでこのexact ground-truth builderの出力contractとして保証する。

## Non-tenpai

34種類すべてのcandidateについてcompletionが存在しなければ、

```text
wait                 = all zero
tanki                = all zero
shanpon              = all zero
kanchan              = all zero
penchan              = all zero
ryanmen_low_side     = all zero
ryanmen_high_side    = all zero
kokushi              = all zero
```

のLevel 2 `HandBelief`を返す。

`None`にはしない。

```text
None     = feature未提供
all-zero = exactにwaitなし
```

というIssue #82の契約を維持する。

## Numerical representation

exact ground truthなのでwait関連tableでは、

```text
0
SCALE
```

以外のraw値を生成しない。

`0.5`等の中間probabilityは生成しない。

expected count / red-five probabilityは、既存 `exact_self_belief()` と同じくconcealed handのみからexact representationを生成する。

## Determinism

同一のsemantic inputに対して必ずbit-exactに同じ `HandBelief` を返す。

以下へ依存してはならない。

- input tile tupleの順序
- set / dict iteration order
- decomposition探索順
- object identity
- hash seed
- environment
- RiichiEnv / RiichiLab

## Implementation strategy

性能最適化よりcorrectnessを優先する。

標準手のdecompositionは34牌種count vectorを使ったrecursive search / memoization等で実装してよい。

ただし、本Issueの目的以上のgeneral-purpose solver frameworkへ過剰抽象化しない。

34 candidate × 小規模hand decompositionで十分なため、profilingなしにRust / NumPy等を導入しない。

## Scope

本Issueでは以下を実装する。

- 完全情報からexact wait ground truthを生成するbuilder
- concealed-onlyのexact tile marginals生成
- own meldsを条件にしたwait導出
- stable 13-equivalent state validation
- structural countとphysical tile countの分離
- 34 candidate tileの決定論的評価
- standard hand completion
- standard handの全成立decompositionを考慮したmechanism分類
- tanki
- shanpon
- kanchan
- penchan
- ryanmen low-side
- ryanmen high-side
- chiitoitsu -> tanki
- kokushi専用channel
- chi / pon / open kan / added kan / concealed kanをcompleted meldとして扱う
- own physical tile count上の5枚目candidate拒否
- malformed physical inputのfail-closed validation
- non-tenpai -> Level 2 all-zero
- exact ground truthの `wait == OR(mechanisms)` 保証
- unit tests
- docs / docstring同期
- 必要なpublic export

## Non-goals

本Issueでは以下を実装しない。

- hidden handからwaitを推定するheuristic
- conditional-uniform beliefへのwait推定
- neural network / learned estimator
- training dataset生成
- 牌譜からtraining sampleを抽出するpipeline
- calibration
- loss function
- Policyへのwait利用
- 放銃危険度
- push / fold
- yaku判定
- scoring
- furiten
- ron / tsumo legality
- global remaining tile inventoryとの結合
- Arena integration
- GameTrace / DecisionTraceへのAI内部情報追加

## Tests

少なくとも以下を固定する。

1. closed standard handのtanki
2. shanpon
3. kanchan
4. `12 -> 3` penchan
5. `89 -> 7` penchan
6. `23 -> 1 / 4` でryanmen low/highを別channelに設定
7. honor tanki
8. honor shanpon
9. chi / ponを含むopen hand
10. 複数meldを持つstandard hand
11. open kanを1 completed meld / 4 physical tilesとして扱う
12. added kanを1 completed meld / 4 physical tilesとして扱う
13. concealed kanを1 completed meld / 4 physical tilesとして扱う
14. concealed kanがあってもchiitoitsuを評価しない
15. concealed kanがあってもkokushiを評価しない
16. 七対子待ちを`tanki`へ設定
17. 七対子を専用channelへ分離しない
18. 国士1種待ち
19. 国士13面待ち
20. 国士を`tanki`へ含めない
21. meld_count > 0ではchiitoitsu / kokushiを評価しない
22. 同一tileに複数standard decompositionが存在するmulti-label case
23. standard + chiitoitsu等、複数hand-family completionが成立する場合のmulti-label
24. non-tenpaiはLevel 2 all-zero
25. wait tableとmechanism ORがexactに一致
26. wait-related raw値が0 / SCALEだけ
27. own concealed hand + meldsですでに4枚持つ牌を5枚目candidateにしない
28. 赤5を通常5と同一basic tile typeとしてphysical countする
29. 入力自体が5枚以上の同一basic tile typeを持つ場合fail-closed
30. 外部remaining inventoryへ依存しない
31. input tile順を変えてもbit-exactに同じ結果
32. invalid 13-equivalent countをfail-closed
33. 14-equivalent drawn stateを非聴牌としてsilentに受理しない
34. meld数 / meld構造がwait判定対象stateとして不正ならfail-closed
35. `expected_count_raw`へmeld tilesを加算しない
36. 既存 `exact_self_belief()` のbehaviorを壊さない
37. Issue #82のLevel 0 / 1 / 2一般contractを壊さない

## Acceptance criteria

- [ ] 完全情報のconcealed hand + own meldsからLevel 2 `HandBelief`を生成できる
- [ ] own meldsはchi / pon / open kan / added kan / concealed kanを扱う
- [ ] tile marginalsはconcealed handのみから生成する
- [ ] wait channelsはconcealed hand + own meldsを条件に生成する
- [ ] すべてのwait-related tableが34牌種canonical axisを使用する
- [ ] stable `len(concealed_tiles) + 3 * meld_count == 13` contractをfail-closedで検証する
- [ ] structural countではkanを1 meld = 3-equivalentとして扱う
- [ ] physical countではkanを4枚として扱う
- [ ] non-tenpaiをall-zero Level 2として表現する
- [ ] standard handの全成立decompositionを考慮する
- [ ] decompositionを恣意的に1つ選ばない
- [ ] tanki / shanpon / kanchan / penchanを正しく分類する
- [ ] ryanmen low-side / high-sideを区別する
- [ ] 七対子は `meld_count == 0` の場合のみ評価し、待ちを`tanki`へ包含する
- [ ] 国士は `meld_count == 0` の場合のみ評価し、専用`kokushi` channelへ設定する
- [ ] own physical tilesだけですでに4枚ある牌について5枚目candidateを拒否する
- [ ] malformedな5枚以上の同一basic tile type入力をfail-closedで拒否する
- [ ] 赤5 / 通常5をphysical count上は同一basic tile typeとして扱う
- [ ] global remaining tile availabilityには依存しない
- [ ] exact outputの各wait raw値は0 / SCALEのみ
- [ ] exact outputでは`wait[t] == OR(mechanisms[t])`を保証する
- [ ] 同一tileのmulti-label mechanismを許容する
- [ ] 同一semantic inputからbit-exactな結果を再現できる
- [ ] RiichiEnv / RiichiLab固有型へ依存しない
- [ ] estimator / Policy / learning pipelineをscopeへ混ぜない
- [ ] 既存 `HandBelief` / `exact_self_belief()` のbehaviorを壊さない
- [ ] docs / docstring / testsが同期する
- [ ] Ruff format / Ruff check / full test suiteが成功する

## 将来の接続

```text
                完全情報牌譜
                     |
                     v
        exact concealed hand + own melds
                     |
                     v
            [今回のIssue]
     exact structural wait builder
                     |
                     v
          HandBelief Level 2
     binary canonical ground truth
                     |
          +----------+----------+
          |                     |
          v                     v
     dataset生成          evaluator / validation
          |
          v
    learned HandBelief
          |
          v
     wait probability
          |
          v
 remaining inventory / value
          |
          v
    deal-in risk / defense
```
